core/mavlink_commands.cpp

```diff
@@ -233,7 +231,15 @@ void MAVLinkCommands::receive_timeout()
     auto work = _work_queue.borrow_front();
 
     if (!work) {
-        // Nevermind, there is nothing to do.
+        // FIXME:
+        // There is nothing to do.
+        // However, we probably shouldn't be in State::WAITING if nothing is in the queue.
+        // Otherwise, we might wait forever because we don't register a new timeout.
```

Comment:
And you observed that we do end here, right? I reviewed the code quickly, I still don't have a full understanding but from what I've seen, it's clear that whenever you set the state to `WAITING`, you register a timeout (which should _always_ be inserted into a map (i.e. I don't see why this insertion would fail)).

The thing I don't understand yet is what `_work_queue.return_front()` does, but I'll have a closer look.

Reply:
Yes, I observed that based on printfs. And once this happened, it was locked up which makes sense because now we're in `WAITING` state when really there is nothing going on and we have to reset to `NONE.

Make sure to also read the commit messages, it might make it clearer.

Correct I'm assuming the registered timeout is always triggered (correctly).

`borrow` leaves the lock locked, so then you need to return it again, so it's unlocked.